File: HB/20200828/정수삼각형.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<vector<int>> triangle) {
    int answer = 0;
    
    int dp[500][500];
    int n = triangle.size();
    
    for(int i = 1; i < n; i++){
        for(int j = 0; j < triangle[i].size(); j++){
            if(j == 0){
                triangle[i][j] = triangle[i][j] + triangle[i-1][j];
            }else if(j == triangle[i].size() - 1){
                triangle[i][j] = triangle[i][j] + triangle[i-1][j-1];
            }else{
                triangle[i][j] += max(triangle[i-1][j], triangle[i-1][j-1]);
            }
        }
    }
    
    answer = *max_element(triangle[n-1].begin(), triangle[n-1].end());
    
    return answer;
}

File: JY/8/8.28/정수삼각형.py
def solution(triangle):
    ans = 0
    n = len(triangle)
    dp = [[0]*i for i in range(1, n+1)]
    
    dp[0][0] = triangle[0][0]
    for i in range(1, n):
        for j in range(i+1):
            if j-1>=0 and j<i:
                dp[i][j] = max(dp[i-1][j-1], dp[i-1][j]) + triangle[i][j]
            elif j==0:
                dp[i][j] = dp[i-1][j] + triangle[i][j]
            elif j==i:
                dp[i][j] = dp[i-1][j-1] + triangle[i][j]
    ans = max(dp[n-1])
    return ans